### DISCUSSION TITLE
error: target(): toolchain not found!

### DISCUSSION THREAD

**Discussion by wsrer** [xmake-io/xmake#3940] [Q&A]
Visual Studio已安装
![image](https://github.com/xmake-io/xmake/assets/69560752/384f0aa9-4bd6-4edf-ae9a-adf79f36465f)
![image](https://github.com/xmake-io/xmake/assets/69560752/b88d2d35-7728-4630-82c5-f953074d1528)
![image](https://github.com/xmake-io/xmake/assets/69560752/338dc095-de93-4ae9-a520-e457f6eb23ea)
![image](https://github.com/xmake-io/xmake/assets/69560752/f77a62d6-fa64-4e32-be04-208b8909294c)
![image](https://github.com/xmake-io/xmake/assets/69560752/060b0ef3-fcfb-4a3e-8c50-04d82ab18c34)
![image](https://github.com/xmake-io/xmake/assets/69560752/e12a5b78-bbe9-4efe-ab35-6058c9b5d684)
![image](https://github.com/xmake-io/xmake/assets/69560752/ecc066c3-7181-45c0-aa12-5eef5bc6fde2)

**Comment by waruqi** [comment]
找到了 vcvarsall.bat ，但是取不到 msvc 的任何信息，通常可能是这里执行失败了。。https://github.com/xmake-io/xmake/blob/710f14b26f29c5d6eaccdf9af197eeca4789ae35/xmake/modules/detect/sdks/find_vstudio.lua#L129

或者没取到返回结果。。

一种可能是 genvcvars_bat 对应的文件路径，写入无权限，比如 c盘加密等情况。。其他可能性，你只能自己调试下。。

可以在上面的脚本位置，你加点 print 后，看下 logs 。。看看失败原因。。

测试命令。

```
xmake g -c
xmake l sdks.detect.find_vstudio
```

也可以打印下 genvcvars_bat 文件路径，然后手动执行下这个 bat 。。看看能否输出取到正确的 envs 内容
